docs/best-practice-consensus-pipeline.md

```diff
@@ -0,0 +1,217 @@
+
+## Tools used
+
+The steps described below use a mixture of:
+
+- `fgbio` (version 2.0 or higher)
+- `samtools` (version 
+- `bwa mem`
+
+
+## Phase 1: FASTQ -> Grouped BAM
+
+The first phase deals with going from FASTQ files through to aligned and grouped BAM files.  The general structure of this phase is:
+
+```mermaid
+graph TD;
+A["fgbio FastqToBam"]-->B["samtools fastq | bwa mem | fgbio ZipperBam"];
+B-->C["fgbio GroupReadsByUmi"];
+```
+
+In this example we'll use a single sample with paired-end fastq files and an inline 8bp UMI at the beginning of read one (see the fgbio [Read Structure](https://github.com/fulcrumgenomics/fgbio/wiki/Read-Structures) page for how to specify other UMI setups).
+
+### Step 1.1: FASTQ -> uBam
+
+When working with fully degenerate UMIs that cannot be error-corrected, the uBam should be generated as follows:
+
+```bash
+fgbio -Xmx1g --compression 1 --async-io FastqToBam \
+  --input r1.fq.gz r2.fq.gz \
+  --read-structures 8M+T +T \
+  --sample s1 \
+  --library library1 \
+  --platform-unit ABCDEAAXX.1 \
+  --output s1.unmapped.bam
+```
+
+If, however, the UMIs are a limited and non-fully degenerate set then they can be error-corrected after extraction from the reads as follows:
+  
+```bash
+fgbio -Xmx1g --compression 0 --async-io FastqToBam \
+  --input r1.fq.gz r2.fq.gz \
+  --read-structures 8M+T +T \
+  --sample s1 \
+  --library library1 \
+  --platform-unit ABCDEAAXX.1 \
+  --output /dev/stdout \
+  | fgbio -Xmx2g --compression 1 CorrectUmis \
+      --input /dev/stdin \
+      --output s1.unmapped.bam \
+      --max-mismatches 1 \
+      --min-distance 2 \
+      --umi-files file-of-expected-umis-one-per-line.txt \
+      --dont-store-original-umis
+```
+
+Neither option above requires or performs any sorting.
+
+### Step 1.2: uBam -> Mapped BAM
+
+This step uses a combination of `samtools` (to convert the uBam back to FASTQ), `bwa mem` to do the actual alignment and `fgbio ZipperBams` to carry metaadata and UMIs forward from the unmapped BAM.  No sorting is performed in this step.
+
+```  
+# Align the data with bwa and recover headers and tags from the unmapped BAM
+samtools fastq s1.unmapped.bam \
+  | bwa mem -t 16 -p -K 150000000 -Y ref.fa \
+  | fgbio -Xmx4g --compression 1 --async-io ZipperBams \
+      --unmapped-bam s1.unmapped.bam \
+      --ref ref.fa \
+      --output s1.mapped.bam 
+```
+
+Some important `bwa mem` options used here are:
+
+- `-t 16` to set the number of threads to use to 16; this can be set to any number between 1 and the number of CPU cores available
+- `-K 150000000` to tell `bwa` to process reads/bases in fixed size chunks regardless of the number of threads; this makes `bwa` deterministic and reproducible with different thread counts
+- `-Y` to use soft-clipping (instead of hard clipping) in supplemntal alignments
```

Comment:
```suggestion
1. `-t 16` to set the number of threads to use to 16; this can be set to any number between 1 and the number of CPU cores available
1. `-K 150000000` to tell `bwa` to process reads/bases in fixed size chunks regardless of the number of threads; this makes `bwa` deterministic and reproducible with different thread counts
1. `-Y` to use soft-clipping (instead of hard clipping) in supplemental alignments; this useful for downstream structural variation detection, among other things
```